File: packages/cli/src/scanner.ts
import { chromium, type Browser, type Page, type Request } from 'playwright';
import {
    VendorRegistry,
    extractDomain,
    isFirstParty,
    type ScanReport,
    type NetworkRequest,
    type DetectedVendor,
    type UnknownDomain,
    type Recommendation,
} from '@etalon/core';

const ETALON_VERSION = '1.0.0';

export interface ScanOptions {
    deep?: boolean;
    timeout?: number;
    waitForNetworkIdle?: boolean;
    userAgent?: string;
    viewport?: { width: number; height: number };
    vendorDbPath?: string;
}

const DEFAULT_OPTIONS: Required<ScanOptions> = {
    deep: false,
    timeout: 30000,
    waitForNetworkIdle: true,
    userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36',
    viewport: { width: 1920, height: 1080 },
    vendorDbPath: '',
};

/**
 * Scan a website for third-party trackers.
 */
export async function scanSite(url: string, options: ScanOptions = {}): Promise<ScanReport> {
    const opts = { ...DEFAULT_OPTIONS, ...options };
    const startTime = Date.now();

    // Load vendor registry
    const registry = opts.vendorDbPath
        ? VendorRegistry.load(opts.vendorDbPath)
        : VendorRegistry.load();

    // Extract site domain for first-party filtering
    const siteDomain = extractDomain(url);
    if (!siteDomain) {
        throw new Error(`Invalid URL: ${url}`);
    }

    const capturedRequests: NetworkRequest[] = [];
    let browser: Browser | null = null;

    try {
        browser = await chromium.launch({ headless: true });
        const context = await browser.newContext({
            userAgent: opts.userAgent,
            viewport: opts.viewport,
            ignoreHTTPSErrors: true,
        });

        const page = await context.newPage();

        // Capture all network requests
        page.on('request', (request: Request) => {
            const reqUrl = request.url();
            const domain = extractDomain(reqUrl);
            if (!domain) return;

            capturedRequests.push({
                url: reqUrl,
                domain,
                method: request.method(),
                type: request.resourceType(),
                timestamp: new Date().toISOString(),
            });
        });

        // Navigate to the page
        await page.goto(url, {
            waitUntil: opts.waitForNetworkIdle ? 'networkidle' : 'domcontentloaded',
            timeout: opts.timeout,
        });

        // Deep scan: scroll, interact, wait for lazy-loaded content
        if (opts.deep) {
            await performDeepScan(page, opts.timeout);
        }

        // Small wait for any final requests
        await page.waitForTimeout(1000);

        await context.close();
    } finally {
        if (browser) {
            await browser.close();
        }
    }

    const scanDurationMs = Date.now() - startTime;

    // Filter to third-party requests only
    const thirdPartyRequests = capturedRequests.filter(
        (req) => !isFirstParty(req.domain, siteDomain)
    );

    // Match against vendor registry
    const vendorMap = new Map<string, DetectedVendor>();
    const unknownMap = new Map<string, UnknownDomain>();

    for (const req of thirdPartyRequests) {
        const vendor = registry.lookupDomain(req.domain);

        if (vendor) {
            const existing = vendorMap.get(vendor.id);
            if (existing) {
                existing.requests.push(req);
            } else {
                vendorMap.set(vendor.id, { vendor, requests: [req] });
            }
        } else {
            const existing = unknownMap.get(req.domain);
            if (existing) {
                existing.requests.push(req);
            } else {
                unknownMap.set(req.domain, {
                    domain: req.domain,
                    requests: [req],
                    suggestedAction: 'submit_for_review',
                });
            }
        }
    }

    const vendors = Array.from(vendorMap.values());
    const unknown = Array.from(unknownMap.values());

    // Sort vendors by risk score (highest first)
    vendors.sort((a, b) => b.vendor.risk_score - a.vendor.risk_score);

    // Generate recommendations
    const recommendations = generateRecommendations(vendors, unknown);

    // Build summary
    const highRisk = vendors.filter((v) => v.vendor.risk_score >= 6).length;
    const mediumRisk = vendors.filter((v) => v.vendor.risk_score >= 3 && v.vendor.risk_score < 6).length;
    const lowRisk = vendors.filter((v) => v.vendor.risk_score < 3).length;

    return {
        meta: {
            etalonVersion: ETALON_VERSION,
            scanDate: new Date().toISOString(),
            scanDurationMs,
            url,
            deep: opts.deep ?? false,
        },
        summary: {
            totalRequests: capturedRequests.length,
            thirdPartyRequests: thirdPartyRequests.length,
            knownVendors: vendors.length,
            unknownDomains: unknown.length,
            highRisk,
            mediumRisk,
            lowRisk,
        },
        vendors,
        unknown,
        recommendations,
    };
}

/**
 * Deep scan: scroll the page, wait for lazy-loaded content.
 */
async function performDeepScan(page: Page, _timeout: number): Promise<void> {
    const scrollDelay = 500;
    const maxScrolls = 10;

    for (let i = 0; i < maxScrolls; i++) {
        await page.evaluate(() => {
            window.scrollBy(0, window.innerHeight);
        });
        await page.waitForTimeout(scrollDelay);
    }

    // Scroll back to top
    await page.evaluate(() => window.scrollTo(0, 0));

    // Wait for any lazy-loaded content
    await page.waitForTimeout(2000);

    // Try to click cookie consent buttons (common patterns)
    const consentSelectors = [
        'button[id*="accept"]',
        'button[id*="consent"]',
        'button[class*="accept"]',
        'button[class*="consent"]',
        'a[id*="accept"]',
        '[data-testid*="accept"]',
        '[data-testid*="consent"]',
    ];

    for (const selector of consentSelectors) {
        try {
            const button = page.locator(selector).first();
            if (await button.isVisible({ timeout: 500 })) {
                await button.click();
                await page.waitForTimeout(1000);
                break;
            }
        } catch {
            // Ignore — consent button not found
        }
    }
}

/**
 * Generate actionable recommendations based on scan results.
 */
function generateRecommendations(
    vendors: DetectedVendor[],
    unknown: UnknownDomain[]
): Recommendation[] {
    const recs: Recommendation[] = [];

    // High-risk vendor alerts
    for (const v of vendors) {
        if (v.vendor.risk_score >= 6) {
            const altText = v.vendor.alternatives?.length
                ? ` Consider alternatives: ${v.vendor.alternatives.join(', ')}`
                : '';
            recs.push({
                type: 'high_risk_vendor',
                vendorId: v.vendor.id,
                message: `${v.vendor.name} is a high-risk tracker (score: ${v.vendor.risk_score}/10).${altText}`,
            });
        }

        // Missing DPA
        if (v.vendor.gdpr_compliant && !v.vendor.dpa_url && v.vendor.risk_score >= 3) {
            recs.push({
                type: 'missing_dpa',
                vendorId: v.vendor.id,
                message: `${v.vendor.name} is GDPR-compliant but no DPA URL is documented. Verify your Data Processing Agreement.`,
            });
        }

        // Suggest alternatives for medium+ risk
        if (v.vendor.risk_score >= 4 && v.vendor.alternatives?.length) {
            recs.push({
                type: 'consider_alternative',
                vendorId: v.vendor.id,
                message: `Consider privacy-friendly alternatives to ${v.vendor.name}: ${v.vendor.alternatives.join(', ')}`,
            });
        }
    }

    // Unknown trackers
    if (unknown.length > 0) {
        recs.push({
            type: 'unknown_tracker',
            message: `${unknown.length} unknown domain(s) detected. Review and submit to the ETALON registry if they are trackers.`,
        });
    }

    return recs;
}

File: crates/cli/src/scanner.rs
use anyhow::Result;
use chromiumoxide::browser::{Browser, BrowserConfig};
use chromiumoxide::cdp::browser_protocol::network::EventRequestWillBeSent;
use etalon_core::domain_utils::{extract_domain, is_first_party};
use etalon_core::registry::VendorRegistry;
use futures::StreamExt;
use std::sync::Arc;
use tokio::sync::Mutex;

#[derive(Debug, Clone)]
pub struct CapturedRequest {
    pub domain: String,
}

pub struct ScanOptions {
    pub deep: bool,
}

pub async fn scan_site(url: &str, options: ScanOptions, registry: &VendorRegistry) -> Result<()> {
    let start_time = std::time::Instant::now();
    tracing::info!("Starting scan for {} (deep={})", url, options.deep);
    let site_domain = extract_domain(url).unwrap_or_else(|| url.to_string());

    // Launch headless browser
    let (mut browser, mut handler) = Browser::launch(
        BrowserConfig::builder()
            .arg("--disable-cache")
            .arg("--incognito")
            .arg("--disable-application-cache")
            .arg("--disable-offline-load-stale-cache")
            .arg("--user-agent=Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36")
            .build()
            .map_err(|e| anyhow::anyhow!(e))?,
    )
    .await?;

    // Spawn handler
    let browser_handle = tokio::spawn(async move {
        while let Some(h) = handler.next().await {
            if h.is_err() {
                break;
            }
        }
    });

    let page = browser.new_page(url).await?;

    // Capture requests
    let captured_requests = Arc::new(Mutex::new(Vec::new()));
    let requests_clone = captured_requests.clone();

    let mut request_events = page.event_listener::<EventRequestWillBeSent>().await?;

    tokio::spawn(async move {
        while let Some(event) = request_events.next().await {
            if let Some(domain) = extract_domain(&event.request.url) {
                let req = CapturedRequest { domain };
                requests_clone.lock().await.push(req);
            }
        }
    });

    // Wait for the page to load or timeout
    let _ = page.wait_for_navigation().await;

    // The key missing piece from Node.js is waiting for network-idle, 
    // where trackers fire *after* the DOM loads. 
    tracing::info!("Waiting for data aggregators and trackers to fire beacons...");

    if options.deep {
        tracing::info!("Deep scanning mode: Scrolling...");
        // Basic deep scan scroll simulation
        for _ in 0..5 {
            let _ = page
                .evaluate("window.scrollBy(0, window.innerHeight);")
                .await;
            tokio::time::sleep(tokio::time::Duration::from_millis(1500)).await;
        }
    } else {
        // Hard wait to allow React/Vue/GTM to bootstrap and fire their analytics
        tokio::time::sleep(tokio::time::Duration::from_millis(5000)).await;
    }

    // Capture requests via Performance API, which catches cross-origin iframe requests
    // and tracker beacons that Chromiumoxide's main-frame event listener misses.
    let js_resources = match page
        .evaluate("Array.from(window.performance.getEntries()).map(e => e.name)")
        .await
    {
        Ok(v) => v.into_value::<Vec<String>>().unwrap_or_default(),
        Err(e) => {
            tracing::warn!("Failed to extract Performance Timing resources: {}", e);
            vec![]
        }
    };
    
    let duration = start_time.elapsed();

    browser.close().await?;
    let _ = browser_handle.await;

    // Process requests
    let mut reqs = captured_requests.lock().await.clone();
    
    for res_url in js_resources {
        if let Some(domain) = extract_domain(&res_url) {
            reqs.push(CapturedRequest { domain });
        }
    }
    let mut third_party_reqs = Vec::new();
    let mut domain_counts: std::collections::HashMap<String, usize> = std::collections::HashMap::new();

    for req in reqs.iter() {
        if !is_first_party(&req.domain, &site_domain) {
            third_party_reqs.push(req.clone());
            *domain_counts.entry(req.domain.clone()).or_insert(0) += 1;
        }
    }

    struct OutputItem {
        domain: String,
        count: usize,
        vendor: Option<etalon_core::types::Vendor>,
    }

    let mut items = Vec::new();
    let mut known_vendors = 0;

    for (domain, count) in domain_counts {
        let vendor = registry.lookup_domain(&domain).cloned();
        if vendor.is_some() {
            known_vendors += 1;
        }
        items.push(OutputItem {
            domain,
            count,
            vendor,
        });
    }

    let mut high_risk = Vec::new();
    let mut medium_risk = Vec::new();
    let mut low_risk = Vec::new();

    for item in items {
        if let Some(v) = &item.vendor {
            if v.risk_score >= 6 {
                high_risk.push(item);
            } else if v.risk_score >= 3 {
                medium_risk.push(item);
            } else {
                low_risk.push(item);
            }
        } else {
            low_risk.push(item); // Unknown defaults to low risk to match legacy behavior
        }
    }

    let dim = "\x1b[90m";
    let cyan = "\x1b[36;1m";
    let red_bold = "\x1b[31;1m";
    let yellow_bold = "\x1b[33;1m";
    let green_bold = "\x1b[32;1m";
    let reset = "\x1b[0m";

    println!("\n\n{}ETALON Privacy Audit{}", cyan, reset);
    println!("{}═══════════════════════════════════════════════════════{}", dim, reset);
    println!("{dim}Site:{reset}       {}", url);
    let now = chrono::Local::now();
    println!("{dim}Scanned:{reset}    {}", now.format("%-d.%-m.%Y, %H:%M:%S"));
    println!("{dim}Duration:{reset}   {:.1} seconds", duration.as_secs_f32());
    println!();

    println!("📊 Summary");
    println!("{}───────────────────────────────────────────────────────{}", dim, reset);
    println!("✓ {} third-party requests", third_party_reqs.len());
    println!("✓ {} matched to known vendors", known_vendors);
    if !high_risk.is_empty() {
        println!("✗ {} high-risk tracker detected", high_risk.len());
    } else {
        println!("✓ 0 high-risk trackers detected");
    }
    println!();

    if !high_risk.is_empty() {
        println!("{}🔴 High Risk ({}){}", red_bold, high_risk.len(), reset);
        println!("{}───────────────────────────────────────────────────────{}", dim, reset);
        for item in high_risk {
            let v = item.vendor.unwrap();
            println!("{:<35} {}", item.domain, v.name);
            println!("{}├─ Category:{reset}   {}", dim, v.category);
            let gdpr_str = if v.gdpr_compliant { "Compliant" } else { "Non-compliant" };
            println!("{}├─ GDPR:{reset}       {}", dim, gdpr_str);
            if let Some(data) = v.data_collected {
                println!("{}├─ Data:{reset}       {}", dim, data.join(", "));
            }
            if let Some(dpa) = v.dpa_url {
                println!("{}├─ DPA:{reset}        {}", dim, dpa);
            }
            if let Some(alts) = v.alternatives {
                println!("{}├─ Alt:{reset}        {}", dim, alts.join(", "));
            }
            println!("{}└─ Requests:{reset}   {}\n", dim, item.count);
        }
    }

    if !medium_risk.is_empty() {
        println!("{}🟡 Medium Risk ({}){}", yellow_bold, medium_risk.len(), reset);
        println!("{}───────────────────────────────────────────────────────{}", dim, reset);
        for item in medium_risk {
            let v = item.vendor.unwrap();
            println!("{:<35} {}", item.domain, v.name);
            println!("{}├─ Category:{reset}   {}", dim, v.category);
            let gdpr_str = if v.gdpr_compliant { "Compliant (with DPA)" } else { "Non-compliant" };
            println!("{}├─ GDPR:{reset}       {}", dim, gdpr_str);
            if let Some(data) = v.data_collected {
                println!("{}├─ Data:{reset}       {}", dim, data.join(", "));
            }
            if let Some(dpa) = v.dpa_url {
                println!("{}├─ DPA:{reset}        {}", dim, dpa);
            }
            if let Some(alts) = v.alternatives {
                println!("{}├─ Alt:{reset}        {}", dim, alts.join(", "));
            }
            println!("{}└─ Requests:{reset}   {}\n", dim, item.count);
        }
    }

    if !low_risk.is_empty() {
        println!("{}🟢 Low Risk ({}){}", green_bold, low_risk.len(), reset);
        println!("{}───────────────────────────────────────────────────────{}", dim, reset);
        for item in low_risk {
            if let Some(v) = item.vendor {
                println!("{:<35} {}", item.domain, v.name);
                println!("{}├─ Category:{reset}   {}", dim, v.category);
            } else {
                println!("{:<35} {}", item.domain, "Unknown Tracker");
            }
            println!("{}└─ Requests:{reset}   {}\n", dim, item.count);
        }
    }

    println!("💡 Recommendations");
    println!("{}───────────────────────────────────────────────────────{}", dim, reset);
    println!("Run with --format json for machine-readable output");
    println!("Report issues: github.com/NMA-vc/etalon/issues\n");

    Ok(())
}
